### ISSUE TITLE
suggestion to remove Process class from scripting

### ISSUE THREAD

**Issue by mrmasterplan** [mapeditor/tiled#3058]
The process class is extremely powerful as it allows arbitrary code execution on the machine running Tiled.

IMHO it is too powerful. Removing the class completely would sandbox any scripting extension to only interact with Tiled. Interactions beyond Tiled should arguably not be initiated by a mapeditor js extension anyway.

Further sandboxing could be achieved by denying extensions from writing files unless a writable file handle is passed to the extension script from Tiled.

**Comment by bjorn** [comment]
It's true that currently scripts have a lot of freedom, essentially the same ones that Tiled itself has. So one needs to be careful when installing scripts from others.

I don't think we can remove Process, since some people do rely on it, and even without process one can execute an arbitrary binary using [`tiled.executeCommand`](https://www.mapeditor.org/docs/scripting/modules/tiled.html#executecommand) (not entirely through scripting API, but the command can be set up in the project, so an attacker would just need to convince somebody to open his Tiled project).

Maybe the Process class should just be disabled by default, so people who rely on it in their project can still enable it. Same probably goes for reading/writing of arbitrary files and XMLHttpRequest. For the files there should indeed probably be some kind of alternative where only files selected by the user can be read/written.

**Comment by mrmasterplan** [comment]
I really like the idea of all the "unsafe" rights being a default-disabled option in the preferences. I'm not bothered by `tiled.executeCommand` at all since executing Tiled functions is what a plugin is supposed to do. Interacting with the rest of the machine, or even the internet, beyond the scope of Tiled is something that I would find suspicious in a scripted extension.

**Comment by bjorn** [comment]
> I'm not bothered by `tiled.executeCommand` at all since executing Tiled functions is what a plugin is supposed to do.

Yeah, I was just saying, if you open somebody else's project in Tiled, this API function is just as dangerous as Process, since somebody could have set up any command in his project for running the desired executable, which an extension installed in the project could then execute through that API.

**Comment by EJlol** [comment]
Microsoft disables macro's when you open up a .docx document from a someone else, and shows a bar with a warning and a button to enable them. Maybe a warning bar like this could also work for commands? This could be done for every project that has commands, or just for projects that are opened on a different computer (however that would also require to store personal data in the project file).

**Comment by bjorn** [comment]
I'm closing this issue since the scripts are not running in a sandbox. I think providing any options beyond the per-project opt-in would suggest a level of security that is hard to guarantee.

Even if we add an option to disable Process or TextFile/BinaryFile, it's hard to make sure scripts really can't use workarounds to still launch processes or read/write files, for example through XMLHttpRequest API (which also allows scripts to access the internet).
